### ISSUE TITLE
[css-fonts-4][css-nesting] Nesting of @supports inside @font-face

### ISSUE THREAD

**Issue by LeaVerou** [w3c/csswg-drafts#6520]
This is related to #633 .

A bit late to the party, but I wondered: what if, instead of adding yet another microsyntax for feature detection, we use `@supports` with an appropriate `font-technology()` function (or whatever we want to name it)?

We [recently resolved](https://github.com/w3c/csswg-drafts/pull/6483) to allow nesting of conditional rules inside regular rules, so that authors can do things like:

```css
a {
	color: red;

	@supports (foo: bar) {
		color: green;
	}
}
```

What if this was allowed in `@font-face` as well? That way, authors could do something like:

```css
@font-face {                                                                                                                                   		
	font-family: Foo;
	src: url("foo.woff2") format("woff2");
	
	@supports font-technology(variations) and font-technology(COLR) and font-technology(palettes) {                                                                                                                     					
		src: url("foo.woff2") format(woff2);
	}
}
```

or 

```css
@font-face {                                                                                                                                   		
	font-family: Foo;
	src: url("foo.woff2") format("woff2");
	
	@supports font-technology(variations COLR palettes) {                                                                                                                     					
		src: url("foo.woff2") format(woff2);
	}
}
```

instead of:

```css
@font-face {                                                                                                                                   		
	font-family: Foo;
	src: url("foo.woff2") format("woff2");                                                                                                                         					
	src: url("foo.woff2") format(woff2 supports variations color(COLR) palettes);                                                               }
```

Benefits:
- It reuses existing syntax, and can be extended to other types of rules in the future, if needed
- It skips the [parsing weirdness for `src`](https://github.com/w3c/csswg-drafts/issues/2920) that this microsyntax has introduced
- It can be used elsewhere, to allow authors to branch their styling differently based on whether these capabilities are allowed. I'd imagine loading a monochrome font over a color font, or a non-variable font would require a fair bit of style differentiation beyond which font is being loaded.
- Authors can decide which descriptors to vary, e.g. they may wish to give the color/variable font a different name
- Arguably easier to read. I find the current proposal quite hard to read, as everything in `format()` is a list of keywords, with no hierarchy. Also, the syntax makes it unclear whether this is a feature query for the browser, or informing the browser what the font file supports.

Downsides:
- I'm hoping UAs can implement this as a one-off, without implementing CSS Nesting in its entirety. My understanding is the functionality is needed _yesterday_, so introducing a big blocker would not be desirable. Would appreciate thoughts from implementors here.
- More verbose

Unlike conditional rules in general, feature queries do not change during the lifetime of the page load, and thus this should not trigger re-interpretation of the `@font-face` rule or be more heavyweight in any other substantial way.

If such syntax is used with today's browsers, they drop the `@supports` rule but keep the `@font-face` rule, so it does appear forwards compatible. [testcase](https://codepen.io/leaverou/pen/eYWapoO)

There are no implementations of the current syntax, so it may not be too late for the change.

Thoughts, @svgeesus @litherum @fantasai @tabatkins?

**Comment by svgeesus** [comment]
This seems obvious but checking just in case: your proposal @LeaVerou  is to allow `@supports` inside **any** at-rule that has descriptors (so, not `@charset` for example), not solely on `@font-face`, right?

**Comment by LeaVerou** [comment]
Currently it is for `@font-face`. Obviously, it would be better to allow it in every at-rule, but I didn't want some unlikely combination of at-rule and `@supports` to hold up this proposal, so I figured we could start small and expand as use cases arise.

**Comment by drott** [comment]
CC @jfkthame

I'll investigate what that would entail from an implementation point of view (complexity, overhead) and comment once I have some findings.

One consideration from a readability point of view: If we introduce such a @supports clause, we will have both mechanisms: a list of entries in the `src:` descriptor (for example multiple `url()` entries potentially then "legacy" `format("...")` specifiers (with its own semantics of choosing the first compatible one from this list)   _and_ potentially multiple `src:` descriptors conditional on `@supports` blocks in the same `@font-face` rule, which I personally do not find great for consistency or readability. In other words, we would change the primary paradigm for `src:` selection to the @supports syntax but would still need to keep compatibility with the current approach of traversing list of entries.

> * It skips the [parsing weirdness for `src`](https://github.com/w3c/csswg-drafts/issues/2920) that this microsyntax has introduced

FWIW, the "parsing weirdness" may be considered addressed or improved with the [most recent change](https://github.com/w3c/csswg-drafts/commit/60dd3ff915b560713b47f4d9b1d0aade6cbb1944) which references the CSS syntax spec for [parsing a comma separated list of components](https://www.w3.org/TR/css-syntax-3/#parse-comma-separated-list-of-component-values).

**Comment by emilio** [comment]
How is the supports rule represented in the OM? Implementation-wise I think it'd be easier for us if it just disappeared, but...

**Comment by LeaVerou** [comment]
Hi @drott!

> One consideration from a readability point of view: If we introduce such a `@supports` rule, we will have both mechanisms: a list of entries in the `src:` descriptor (for example multiple `url()` entries potentially then "legacy" `format("...")` specifiers (with its own semantics of choosing the first compatible one from this list) _and_ potentially multiple `src:` descriptors conditional on `@supports` blocks in the same `@font-face` rule, which I personally do not find great for consistency or readability. In other words, we would change the primary paradigm for `src:` selection to the `@supports` syntax but would still need to keep compatibility with the current approach of traversing a list of entries.

Yeah, this was raised by @plinss as well. The counter-argument he agreed to was that `@supports` came after `format()`, so while we're stuck with `format()` for legacy format selection, we shouldn't be *actively* developing two different feature detection mechanisms, one of which is entirely ad hoc and severely more limited.

> FWIW, the "parsing weirdness" may be considered addressed or improved with the [most recent change](https://github.com/w3c/csswg-drafts/commit/60dd3ff915b560713b47f4d9b1d0aade6cbb1944) which references the CSS syntax spec for [parsing a comma separated list of components](https://www.w3.org/TR/css-syntax-3/#parse-comma-separated-list-of-component-values).

I was referring to the fact that this descriptor's syntax needs prose and cannot be fully described by a grammar. I wasn't even aware that it used to be *even* weirder, whoa!

@emilio

> How is the supports rule represented in the OM? Implementation-wise I think it'd be easier for us if it just disappeared, but...

I've wondered about this too, but I'm afraid making it disappear would be fairly inconsistent with every other @-rule, which would be even more confusing once Nesting is implemented.

**Comment by tabatkins** [comment]
While it would involve more duplication, would it be okay to just leverage @supports normally, with some new queries to ask about various font-feature support, and then just have @font-face as a child of the rule like normal?

The big argument against is if we think multiple independent queries affecting different bits would be common; you'd have a combinatorial explosion to achieve equivalent results.

**Comment by tabatkins** [comment]
Looking at the proposal itself - it feels *slightly* weird to have context-specific @supports (I presume this *wouldn't* allow arbitrary property-support queries, but only font-based ones?). If I'm wrong about that and this would just be an ordinary @supports, with a supporting proposal to add the font-feature queries, then this is less weird. (If I'm right, naming it `@font-supports` or something would defuse my objection.)

I don't have a *strong* opinion on whether this should be treated as a parse-time syntax, disappearing from the tree in the OM, or kept around like normal @supports is. Parse-time is definitely simpler overall, but I understand how it feels new and odd.

Overall tho, I definitely support *something* like this over the growing weird microsyntax. The microsyntax was reasonable when it was just one thing, but it's getting out of hand.

**Comment by LeaVerou** [comment]
> While it would involve more duplication, would it be okay to just leverage @supports normally, with some new queries to ask about various font-feature support, and then just have @font-face as a child of the rule like normal?
> 
> The big argument against is if we think multiple independent queries affecting different bits would be common; you'd have a combinatorial explosion to achieve equivalent results.

I think that could be a good first step, and would enable authors to do what they want with some duplication. However, we do eventually want to avoid the duplication, even if multiple independent queries aren't common.

> Looking at the proposal itself - it feels _slightly_ weird to have context-specific @supports (I presume this _wouldn't_ allow arbitrary property-support queries, but only font-based ones?). If I'm wrong about that and this would just be an ordinary @supports, with a supporting proposal to add the font-feature queries, then this is less weird. (If I'm right, naming it `@font-supports` or something would defuse my objection.)

It is not context-specific. The whole idea is that this way authors can use it to feature detect in any context that suits them, not just to differentiate the `src`, but possibly other aspects of their style.

**Comment by drott** [comment]
Having looked at and discussed with my CSS colleagues, the implementation of `@supports` inside the `@font-face` at-rule, would be a larger effort, in particular if the `@supports` rules need to be represented in the CSSOM, which poses problems given the current internal representation. (In addition, we're blocked on some parser refactoring which would make this more difficult, but that's an internal issue).

>> While it would involve more duplication, would it be okay to just leverage @supports normally, with some new queries to ask about various font-feature support, and then just have @font-face as a child of the rule like normal?

> I think that could be a good first step, and would enable authors to do what they want with some duplication, and would likely (?) be relatively easy to implement. However, we do eventually want to avoid the duplication, even if multiple independent queries aren't common.

Yes, I agree, adding something similar to a `<supports-font-technology-fn>` to **[2. Extensions to the @supports rule](https://www.w3.org/TR/css-conditional-4/#at-supports-ext)** with syntax similar to what is in the current font spec proposal is immediately useful for the desired feature detection through `CSS.supports()` JavaScript and would immediately work at the top level (not nested).

This would give us a path that's relatively straight-forward to implement and not blocked on nesting `@supports` inside `@font-face`, while it would still allow for that kind of nesting later and migrating away from unnecessary `@font-face` repetition.

**Comment by LeaVerou** [comment]
I think adding the `@supports` criteria first, and implementing the nesting step later is a good path forwards.

Adding this to the agenda so we can resolve that these changes have CSS WG consensus.

**Comment by drott** [comment]
In [this branch](https://github.com/drott/csswg-drafts/commit/7a63d1c44c98c4971d2922dce8094dcb48d44b7a) I drafted an edit to [css-conditional-4](https://drafts.csswg.org/css-conditional-4/#at-supports-ext), carrying over the syntax from [fonts-4's - Parsing the source descriptor](https://drafts.csswg.org/css-fonts/#font-face-src-parsing), an excerpt from that draft (feel free to comment on the [commit](https://github.com/drott/csswg-drafts/commit/7a63d1c44c98c4971d2922dce8094dcb48d44b7a) directly as well):

```
<supports-feature> = <supports-selector-fn> | <supports-font-technology-fn> | <supports-decl>
<supports-selector-fn> = selector( <complex-selector> )
<supports-font-technology-fn> = font-technology ( <font-technology> )
<font-technology> = [ features( <font-feature-technology> ) | variations
                    | color( <color-font-technology> ) | palettes | incremental ]
<font-feature-technology> = [ opentype | aat | graphite]
<color-font-technology> = [ COLRv0 | COLRv1 | SVG | sbix | CBDT ]
```

Considerations:

* I did not add a repetition operator _#_ after `<font-technology>` in the `<supports-font-technology-fn>` rule as required combinations of technologies can be expressed through `and` and `or` of the `@supports` syntax, which I consider an improvement over the syntax in the fonts spec.

I'd be very happy if in tomorrow's meeting we could attempt to resolve towards a direction of the syntax along those lines (feedback very welcome), in addition to the general aim to resolve adding the criteria to the `@supports` rule.

**Comment by svgeesus** [comment]
> I did not add a repetition operator # after <font-technology> in the <supports-font-technology-fn> rule as required combinations of technologies can be expressed through and and or of the @supports syntax, which I consider an improvement over the syntax in the fonts spec.

I agree, more readable that way.

**Comment by svgeesus** [comment]
Just to be sure I understand, the first example in [the`@font-face` `src` explainer](https://github.com/w3c/csswg-drafts/blob/main/css-fonts-4/src-explainer.md) would become (repetitive, but readable; most desirable option is last):

```css
/* 1. prefer COLRv1, then SVG-in-OpenType, then COLRv0 */
@font-face {
  font-family: jewel;
  src: url(boring.ttf) format("woff2");
}
@supports  font-technology(color(COLRv0)) {
    @font-face {
        font-family: jewel;
        src: url(jewel-v0.woff2) format("woff2");
    }
}
@supports  font-technology(color(SVG)) {
    @font-face {
        font-family: jewel;
        src: url(jewel-svg.woff2) format("woff2");
    }
}
@supports  font-technology(color(COLRv1)) {
    @font-face {
        font-family: jewel;
        src: url(jewel-v1.woff2) format("woff2"),;
    }
}
```

**Comment by drott** [comment]
Yes, that's how I would express it as well. That is: for now, without nested `@supports`.

**Comment by svgeesus** [comment]
@dbaron @fantasai thoughts? Since we are all editors of [Conditional 3](https://drafts.csswg.org/css-conditional-3/) I assume we are of [Conditional 4 too](https://drafts.csswg.org/css-conditional-4/), right? 
@astearns @atanassov or would that need a separate resolution?

**Comment by svgeesus** [comment]
Oh wait

> The declaration being tested must always occur within parentheses, when it is the only thing in the expression. [source](https://drafts.csswg.org/css-conditional-3/#at-supports)

So that is like

```css
@supports  (font-technology(color(COLRv0)) ){
and so on
```

**Comment by LeaVerou** [comment]
> Oh wait
> 
> > The declaration being tested must always occur within parentheses, when it is the only thing in the expression. [source](https://drafts.csswg.org/css-conditional-3/#at-supports)
> 
> So that is like
> 
> ```css
> @supports  (font-technology(color(COLRv0))) {
>    and so on
> ```

This applies to declaration tests, not `selector()` or `font-technology()` queries.

Nitpick, I don’t think we need nested function calls, not to mention `color()` might look confusing because of the other `color()` function. We could just prefix keywords with `color-` and `feature-`, which avoids the multiple nested parens.

**Comment by drott** [comment]
> We could just prefix keywords with `color-` and `feature-`, which avoids the multiple nested parens.

Updated the proposed change with that suggestion, thanks.

**Comment by LeaVerou** [comment]
@svgeesus asked yesterday if we could use colons (`font-technology(color: COLRv0)`). I *think* that's not idiomatic to CSS and it's weird to have key-value pairs for some things and not others, but I'd defer to @tabatkins on that one. If he's ok with colons, I’m ok with colons.

**Comment by svgeesus** [comment]
So, that would be

```css
/* 1. prefer COLRv1, then SVG-in-OpenType, then COLRv0 */
@font-face {
  font-family: jewel;
  src: url(boring.ttf) format("woff2");
}
@supports  font-technology(color-COLRv0) {
    @font-face {
        font-family: jewel;
        src: url(jewel-v0.woff2) format("woff2");
    }
}
@supports  font-technology(color-SVG) {
    @font-face {
        font-family: jewel;
        src: url(jewel-svg.woff2) format("woff2");
    }
}
@supports  font-technology(color-COLRv1) {
    @font-face {
        font-family: jewel;
        src: url(jewel-v1.woff2) format("woff2"),;
    }
}
```

**Comment by svgeesus** [comment]
>  it's weird to have key-value pairs for some things and not others

Its also weird to have key-value pairs multiple times for the same key

```css
@supports font-technology(color: COLRv0) AND font-technology(color: COLRv1)
```

so on balance I prefer the hyphenated form proposed earlier.

**Comment by drott** [comment]
> The CSS spec even suggests that it's reasonable to use broad-range unicode-range descriptors rather than enumerating the precise coverage of the font:

I'd be in favour of changing that recommendation.

In the discussion so far, I see these constraints - am I missing something, or are there differing views?

1. Due to implementation complexity and CSSOM implications it's harder to put `@supports` inside `@font-face` so we seem to agree on a preference to have it at the top level. 
2. If `@support` + a `font-technology()` function is placed at the top level, we may run into the cascading / accumulation issue (with varying opinions on how sever that issue is). 
3. If we define a `font-technology()` supports function, nothing per-se stops it from being used at the top-level. 

Options we've been pondering in an internal discussion:

- Add a new descriptor to `@font-face` remotely similar to a C++ `final` keyword - meaning that there shouldn't be further search for other segmented font faces in the same family. This would require people to use it, otherwise the same cascading problem remains. And due to the fact it needs to be actively used it becomes somewhat redundant with `unicode-range` which addresses the same issue. Alternatively, but probably too intrusive, have a new `@font-face-no-cascade` (name to be bikeshod) with a different matching algorithm / cascading behavior?
- Would it be useful to define something like a `url()` alias blob object to point to a `url()` resource inside a top-level `@supports` - then use that alias/reference in a `src:` descriptor inside `@font-face`s? Are there precedents for something like that?
- If we don't want `@supports` at the top level and despite the implementation hurdles opt for a nested approach, add prose to the new `font-technology()` function that it resolves to false when used outside the nested scope?

**Comment by tabatkins** [comment]
I apologize, but it's not entirely clear to me what this `final` concept would bring that's not already done by omitting `unicode-range` entirely.

**Comment by drott** [comment]
The way I mean it, taking [Jonathan's example](https://github.com/w3c/csswg-drafts/issues/6520#issuecomment-906384393) and adding a hypothetical `fallback: no-cascade; `

```
@font-face { font-family: my-icons; src: url(my-icons-bw-fallback.woff2); fallback: no-cascade; }

@supports font-technology(color-COLRv0) {
  @font-face { font-family: my-icons; src: url(my-icons-flat-colors.woff2); fallback: no-cascade; }
}

@supports font-technology(color-COLRv1) {
  @font-face { font-family: my-icons; src: url(my-icons-gradient-colors.woff2); fallback: no-cascade; }
}
```

Without `fallback: no-cascade;` the _my-icons_ family would consist of three fonts that have accumulated into the _my-icons_ font at identical attributes for weight, stretch, style. So when checking for a character that's not contained in the icon font, first _my-icons_ COLRv1, then COLRv0, then bw have to be loaded, in that order. If there is a hyptohtetical `fallback: no-cascade;` descriptor, we can spec that to mean: after checking the first one (in reverse order of CSS declarations), do not look further in this cascade and do not load any others.

The way I understand it: Omitting `unicode-range` would usually lead to all of them being loaded and checked, until system fallback is used. Using `unicode-range` with tight codepoint ranges means a behavior similar to the hypothetical `fallback: no-cascade;`, meaning: if `unicode-range` indicates there's not going to be coverage, do not load this font.

**Comment by drott** [comment]
Yes, the syntax you drafted resembles very closely the proposal that @tabatkins previously published here:
https://tabatkins.github.io/specs/css-when-else/ which was [briefly mentioned in the notes](https://github.com/w3c/csswg-drafts/issues/6520#issuecomment-905712330) for the last CSS WG meeting. 

Yes, perhaps if we explore this further we'll have a future ergonomics improvement for `@supports` & `font-technology()`. We could define the latter now and invest in `@when + @else` - which might be a viable way forward.

**Comment by litherum** [comment]
Pursuing `@else` would satisfy my concerns, too. `@else` would have to come first, though, so we don't teach authors to do the wrong thing because the better tools aren't available yet.

If we have `@else`, I don't see any reason why other at-rules couldn't be inside it, too. `@font-face` doesn't need to be special.

**Comment by svgeesus** [comment]
Relevant issue for an `if-elsif-else` block is https://github.com/w3c/csswg-drafts/issues/112

I agree that we need a syntax where: 

- exactly one branch is chosen (finding a winning one means you don't evaluate the rest)
- the logical tests for each branch can combine multiple boolean tests

Note that the existing, ugly, jammed-into-format syntax already has those properties; and I don't want to see it adopted anyway just for expediency because we take too long to add an `if-elsif-else` construct. I also don't want us to adopt a solution "for now" that has the multiple download and fallback problems that @jfkthame has explained so well.

> My only remaining thought is general distaste for using different mechanisms to select truetype/opentype vs COLR/sbix.

We will need to keep `format` for legacy reasons but nothing prevents our adding a format test to the new syntax along with the font technology tests. Although the spec already requires treating truetype and opentype as synonyms so `format` is mainly testing for new compression/deployment wrappers (woff, woff2, IFT, etc)

**Comment by drott** [comment]
@litherum wrote:

> @else would have to come first, though, so we don't teach authors to do the wrong thing because the better tools aren't available yet.

@svgeesus wrote:

> I also don't want us to adopt a solution "for now" that has the multiple download and fallback problems that @jfkthame has explained so well.

If we add an `@else` construct, and only _after that_ define a `font-technology()` function for feature detection and `@supports` conditionals, we supposedly gain a form of educational advantage.

But if both are shipped in the future, whom are we preventing from still writing the less performance optimised version that @jfkthame explained where the font blobs accumulated into a family through which we need to fallback?

Aren't we overestimating this advantage of shipping @else before the feature detection function, given that it will still be possible to spell out the less performant version (which I still consider a niche case given that font CSS is generated in the majority of use cases)? Aren't we undererstimating the uptake of a better `@else` construct once it's available?

**Comment by litherum** [comment]
I added 2 comments to https://github.com/drott/csswg-drafts/commit/62cfd95a5f52604c952e4aa37be6e4d6386f88f5 asking for a note to implementors.

**Comment by svgeesus** [comment]
@drott I'm waiting for this commit to surface as a PR on conditional 4, so that I can then merge in Tab's proposal without a merge conflict

**Comment by svgeesus** [comment]
> I'm waiting for this commit to surface as a PR on conditional 4, so that I can then merge in Tab's proposal without a merge conflict

Both now done.

**Comment by svgeesus** [comment]
>  It also might be worth saying that implementing this depends on implementing @else.

I added a pair of examples that shows how `@else` is a cleaner solution for conditional `@font-face`

**Comment by fantasai** [comment]
Posting what I think we should do here.

Considerations:
- Font tech features are dependent on the font format. We do have multiple formats which are effectively interchangeable wrappers right now, and for those the wrapper format and feature tech are independent queries, but this isn't true in consideration of all font formats past and future.
- Authors benefit from consistent syntax. If we have a query syntax in both `src` and `@supports`, they should be the same.
- Because `@font-face` rules combine as they cascade, fallback code needs to actually sit in its own conditional; we can't rely on later rules overwriting earlier ones.
- The `format()` micro-query syntax with nested functions that was proposed earlier was too complex.

Proposal:
- Extend the `format()` function to take keywords after the format string. These keywords represent the various font tech features.
- Extend `@supports` to add a `font-format()` function with identical syntax.
- Add `@else` to allow fallback rules to be easily made conditional on negation of the `@supports` query.

```
<font-format> = <format-string> <ident>*
<format-string> = format type keyword registered in css-fonts or any registered MIME type keyword under the `font/` hierarchy

src: <url> format(<font-format>), ...;
@supports font-format(<font-format>) { ... }
```

**Comment by fantasai** [comment]
> @fantasai with that simplification src won't need special parsing rules, right?

Correct.

**Comment by fantasai** [comment]
Agenda+ to discuss https://github.com/w3c/csswg-drafts/issues/6520#issuecomment-934811808 and whether/what subset we want to approve for shipping (since Chrome wants to ship something soon).

**Comment by svgeesus** [comment]
> Font tech features are dependent on the font format. We do have multiple formats which are effectively interchangeable wrappers right now, and for those the wrapper format and feature tech are independent queries, but this isn't true in consideration of all font formats past and future.

That isn't really true currently, which makes it a poor fit for `format`. 

Agreed [in the past](https://www.w3.org/TR/2002/WD-css3-webfonts-20020802/#referencing) there were

- "speedo"
- "intellifont"
- "truedoc-pfr" 
- "SVG"

and of course we can't predict what may happen more than a few years into the future, but currently in font format terms everything is an `sfnt` which could be called either TrueType or OpenType. And (also stuffed into format) there may be compression/transfer tech on top (WOFF|WOFF2|IFT) but that does not alter the font technologies available, it just compresses or subsets them.

That said I don't object to renaming `font-technology()` to `font-format()` if that is what works and we agree on. But it makes things less clear.

**Comment by svgeesus** [comment]
We would also need to test what happens with

```css
@font-format {
  font-family: foo;
  src: url("foo.woff") format("color-COLRv1");
}
```
because I suspect with an omitted actual format, the font is assumed to be an unsupported format and is not loaded.

In general the whole `format()` thing is a buggy hack with huge backwards-compatible fragility; adding anything is highly constrained and we can't do good design in this space. I would be much happier with (as we originally resolved)

1. dropping [supports &lt;font-technology>#]? from `@font-face` `src` entirely
2. using the conditional features in Conditional 4 instead, together with `font-technology()`

**Comment by svgeesus** [comment]
Within the comma separated list, things are supposed to be space separated. Which is why

```css
format(opentype supports incremental)
```

works in a backward compat way. As you say, older browsers treat the whole thing as one unknown format that happens to contain spaces.

**Comment by LeaVerou** [comment]
@litherum My understanding of the [`@supports` grammar](https://drafts.csswg.org/css-conditional-3/#at-supports) is that any negated condition needs to be in parens. If that's correct, that would explain why your first rule didn't work: because it's invalid syntax. Perhaps @tabatkins could confirm.

**Comment by astearns** [comment]
Full minutes posted here: https://www.w3.org/2021/10/20-css-minutes.html

**Comment by svgeesus** [comment]
> I thought we could get rid of the parsing weirdness around src now?

We got rid of the weirdness around `format supports whatever`. We still have a descriptor wchich is a comma-separated list of items, where each item is a `url()` followed by an optional `format()` and an optional `technology()` so we still need to describe how to work out the winning item.
